File: SGUEES-SPA/src/app/pages/Payroll/pla-seguro-social/pla-seguro-social.service.ts
// Qué hace: servicio de negocio del catálogo Seguro Social.
// Cómo lo hace: valida, ejecuta CRUD/ActivarInactivar y arma columnas/formulario.
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { IParam } from 'src/app/FxAPI/IParam';
import { IResult } from 'src/app/FxAPI/IResult';
import { NotifyType } from 'src/app/shared/models/NotifyType';
import { buildAuditGridColumns } from 'src/app/shared/mtto/mtto-grid.helpers';
import { createEstadoColumnConfig, ESTADO_ACTIVO_INACTIVO_LABELS } from 'src/app/shared/utils/remote-grid-filter.util';
import { PlaSeguroSocial } from './models/pla-seguro-social';
import { PlaSeguroSocialRepository } from './pla-seguro-social.repository';

const ESTADO_FIELD = 'ACTIVO_SEGURO_SOCIAL';

@Injectable({ providedIn: 'root' })
export class PlaSeguroSocialService {
	constructor(private repo: PlaSeguroSocialRepository) {}

	esValido(model: PlaSeguroSocial, msg: Function): boolean {
		if (!model.NOMBRE_SEGURO_SOCIAL || model.NOMBRE_SEGURO_SOCIAL.trim() === '') {
			msg('Debe ingresar el nombre del seguro social.', NotifyType.Warning);
			return false;
		}
		if (model.NOMBRE_SEGURO_SOCIAL.trim().length > 150) {
			msg('El nombre no puede superar 150 caracteres.', NotifyType.Warning);
			return false;
		}
		if (!model.NOMBRE_CORTO_SEGURO || model.NOMBRE_CORTO_SEGURO.trim() === '') {
			msg('Debe ingresar el nombre corto del seguro social.', NotifyType.Warning);
			return false;
		}
		if (model.NOMBRE_CORTO_SEGURO.trim().length > 50) {
			msg('El nombre corto no puede superar 50 caracteres.', NotifyType.Warning);
			return false;
		}
		return true;
	}

	getAll(param: any): Observable<IResult> {
		return this.repo.getAll(this.buildWhere(param));
	}

	get(param: any): Observable<IResult> {
		return this.repo.get([{ Parameter: 'CORR_SEGURO_SOCIAL', Value: param.CORR_SEGURO_SOCIAL }]);
	}

	insert(model: any): Observable<IResult> {
		return this.repo.create(model);
	}

	update(model: any): Observable<IResult> {
		return this.repo.update(model, [{ Parameter: 'CORR_SEGURO_SOCIAL', Value: model.CORR_SEGURO_SOCIAL }]);
	}

	delete(model: any): Observable<IResult> {
		return this.repo.delete([{ Parameter: 'CORR_SEGURO_SOCIAL', Value: model.CORR_SEGURO_SOCIAL }]);
	}

	activarInactivar(model: any): Observable<IResult> {
		return this.repo.activarInactivar(model, [
			{ Parameter: 'CORR_SEGURO_SOCIAL', Value: model.CORR_SEGURO_SOCIAL },
		]);
	}

	getColumns(): any {
		return [
			{
				dataField: 'CORR_SEGURO_SOCIAL',
				caption: 'Corr.',
				width: 90,
				dataType: 'number',
				filterOperations: ['=', '<', '>', '<=', '>='],
			},
			{ dataField: 'NOMBRE_SEGURO_SOCIAL', caption: 'Nombre seguro social', width: 300, minWidth: 200 },
			{ dataField: 'NOMBRE_CORTO_SEGURO', caption: 'Nombre corto', width: 180, minWidth: 120 },
			createEstadoColumnConfig(ESTADO_FIELD, ESTADO_ACTIVO_INACTIVO_LABELS, { caption: 'Estado' }),
			...buildAuditGridColumns({ withDateTimeFilter: true }),
		];
	}

	getSummary(): any {
		return {
			totalItems: [
				{
					column: 'CORR_SEGURO_SOCIAL',
					summaryType: 'count',
					valueFormat: '#,##0',
					displayFormat: 'Cant: {0}',
				},
			],
		};
	}

	getItems(): any {
		return [
			{ dataField: 'CORR_SEGURO_SOCIAL', label: { text: 'Corr.' }, colSpan: 1, editorOptions: { readOnly: true } },
			{
				dataField: 'NOMBRE_SEGURO_SOCIAL',
				label: { text: 'Nombre seguro social' },
				colSpan: 5,
				editorOptions: { placeholder: 'Nombre seguro social...', showClearButton: true, maxLength: 150 },
				validationRules: [{ type: 'required', message: 'Este campo es obligatorio' }],
			},
			{
				dataField: 'NOMBRE_CORTO_SEGURO',
				label: { text: 'Nombre corto' },
				colSpan: 4,
				editorOptions: { placeholder: 'Nombre corto...', showClearButton: true, maxLength: 50 },
				validationRules: [{ type: 'required', message: 'Este campo es obligatorio' }],
			},
			{ dataField: 'ACTIVO_SEGURO_SOCIAL', label: { text: 'Activo' }, editorType: 'dxCheckBox', colSpan: 2 },
		];
	}

	private buildWhere(param: any): IParam[] {
		const xWhere: IParam[] = [];
		if (param.CORR_SEGURO_SOCIAL) {
			xWhere.push({ Parameter: 'CORR_SEGURO_SOCIAL', Value: param.CORR_SEGURO_SOCIAL });
		}
		return xWhere;
	}
}

File: SGUEES-API/sguees.api/Options/Payroll/PLA_SEGURO_SOCIAL/PLA_SEGURO_SOCIALService.cs
// Qué hace: capa de servicio del catálogo Seguro Social.
// Cómo lo hace: valida datos y delega CRUD/ActivarInactivar al repositorio.
using System.Collections.Generic;
using System.Threading.Tasks;
using eFramework.Core;
using eFramework.Data;
using SGUEES.Models;
using SGUEES.Repositories;

namespace SGUEES.Services
{
	public class PLA_SEGURO_SOCIALService : IPLA_SEGURO_SOCIALService
	{
		private readonly IPLA_SEGURO_SOCIALRepository _repo;

		public PLA_SEGURO_SOCIALService(IPLA_SEGURO_SOCIALRepository repo)
		{
			_repo = repo;
		}

		public async Task<CResult> GetAllAsync(PLA_SEGURO_SOCIALParam xWhere)
		{
			return await _repo.GetAllAsync(new List<CParameter>());
		}

		public async Task<CResult> GetAsync(PLA_SEGURO_SOCIALParam xWhere)
		{
			var p = new List<CParameter>
			{
				new CParameter() { ParameterName = "CORR_SEGURO_SOCIAL", Value = xWhere.CORR_SEGURO_SOCIAL, DbType = System.Data.DbType.Int32 },
			};

			return await _repo.GetAsync(p);
		}

		public async Task<CResult> CreateAsync(PLA_SEGURO_SOCIALTable Data, string vLOGIN_SISTEMA, string vESTACION)
		{
			var validation = Validate(Data);
			if (validation != null)
			{
				return validation;
			}

			NormalizeData(Data);
			return await _repo.CreateAsync(Data, vLOGIN_SISTEMA, vESTACION);
		}

		public async Task<CResult> UpdateAsync(PLA_SEGURO_SOCIALTable Data, string vLOGIN_SISTEMA, string vESTACION)
		{
			var validation = Validate(Data);
			if (validation != null)
			{
				return validation;
			}

			if (Data.CORR_SEGURO_SOCIAL <= 0)
			{
				return ValidationError("No se pudo identificar el seguro social a actualizar.");
			}

			NormalizeData(Data);
			return await _repo.UpdateAsync(Data, vLOGIN_SISTEMA, vESTACION);
		}

		public async Task<CResult> DeleteAsync(PLA_SEGURO_SOCIALTable Data, string vLOGIN_SISTEMA, string vESTACION)
		{
			return await _repo.DeleteAsync(Data, vLOGIN_SISTEMA, vESTACION);
		}

		public async Task<CResult> ActivarInactivarAsync(PLA_SEGURO_SOCIALTable Data, string vLOGIN_SISTEMA, string vESTACION)
		{
			if (Data.CORR_SEGURO_SOCIAL <= 0)
			{
				return ValidationError("No se pudo identificar el seguro social a actualizar.");
			}

			return await _repo.ActivarInactivarAsync(Data, vLOGIN_SISTEMA, vESTACION);
		}

		private static void NormalizeData(PLA_SEGURO_SOCIALTable Data)
		{
			Data.NOMBRE_SEGURO_SOCIAL = (Data.NOMBRE_SEGURO_SOCIAL ?? string.Empty).Trim();
			Data.NOMBRE_CORTO_SEGURO = (Data.NOMBRE_CORTO_SEGURO ?? string.Empty).Trim();
			Data.ACTIVO_SEGURO_SOCIAL ??= true;
		}

		private static CResult Validate(PLA_SEGURO_SOCIALTable Data)
		{
			if (string.IsNullOrWhiteSpace(Data.NOMBRE_SEGURO_SOCIAL))
			{
				return ValidationError("Debe ingresar el nombre del seguro social.");
			}

			if (Data.NOMBRE_SEGURO_SOCIAL.Trim().Length > 150)
			{
				return ValidationError("El nombre del seguro social no puede superar 150 caracteres.");
			}

			if (string.IsNullOrWhiteSpace(Data.NOMBRE_CORTO_SEGURO))
			{
				return ValidationError("Debe ingresar el nombre corto del seguro social.");
			}

			if (Data.NOMBRE_CORTO_SEGURO.Trim().Length > 50)
			{
				return ValidationError("El nombre corto no puede superar 50 caracteres.");
			}

			return null;
		}

		private static CResult ValidationError(string message)
		{
			return new CResult
			{
				Result = false,
				ErrorCode = 400,
				ErrorMessage = message,
			};
		}
	}
}
